framework/rbac/DbManager.php

```diff
@@ -501,6 +501,39 @@ public function getPermissionsByUser($userId)
             return [];
         }
 
+        $directPermission = $this->getDirectPermissionsByUser($userId);
+        $inheritedPermission = $this->getInheritedPermissionsByUser($userId);
+
+        return array_merge($directPermission, $inheritedPermission);
+    }
+
+    /**
+     * Returns all permissions that are directly assigned to user.
+     * @param string|integer $userId the user ID (see [[\yii\web\User::id]])
+     * @return Permission[] all direct permissions that the user has. The array is indexed by the permission names.
+     */
+    protected function getDirectPermissionsByUser($userId)
```

Comment:
any reason why this method is protected? It would've been easier to implement a form for changing the permissions too.

Reply:
Well it's been three years and I don't recall the details, but the problem was that the `getPermissionsByUser()` method did not return directly assigned permissions, only returning permissions connected to a user via assigned roles.

Suggested solution aimed to resolve that issue, having `getDirectPermissionsByUser()` as public method would've been a new feature (maybe a useful one) that I'm not sure if maintainers would agree with at the time.

And I don't get the form suggestion in this context. can you elaborate please?

Reply:
Thanks for the explanation. We have implemented a profile page for users, where admins can assign roles to users. We also added the possibility to assign extra permissions to users, for a more granular access schema, for example giving access to team members from another team just to view items, but not edit them.

Problem is, if we use `getPermissionsByUser()` we fill the extra permissions input with the inherited ones too, so on second form submit, all those will be directly assigned to user.

Right now I extended the DbManager class to make `getDirectPermissionsByUser()` public so we can only display and edit those ones in the form.

I can open a PR if maintainers have nothing against.

Reply:
We had the exact same problem! and made a workaround just like you. didn't cross my mind to make these methods public in framework's core though. 
please let me know when you created the new issue, it'll have a thumb up from me :)